src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1789,6 +1954,120 @@ Node TheoryStringsRewriter::collectConstantStringAt( std::vector< Node >& vec, u
   }
 }
 
+bool TheoryStringsRewriter::stripSymbolicLength(std::vector<Node>& n1,
+                                                std::vector<Node>& nr,
+                                                int dir,
+                                                Node& curr)
+{
+  Assert(dir == 1 || dir == -1);
+  Assert(nr.empty());
+  Node zero = NodeManager::currentNM()->mkConst(CVC4::Rational(0));
+  bool ret = false;
+  bool success;
+  unsigned sindex = 0;
+  do
+  {
+    Assert(!curr.isNull());
+    success = false;
+    if (curr != zero && sindex < n1.size())
+    {
+      unsigned sindex_use = dir == 1 ? sindex : ((n1.size() - 1) - sindex);
```

Comment:
For vectors, an alternative to this pattern could be to use either an iterator created with `begin()` or a reverse iterator created with `rbegin()` depending on dir. This could simplify some of the code dealing with the direction in this function.

Reply:
I like having sindex as an unsigned here.  I have been toying with the idea of making the return value of this function equal to the resulting value of sindex, analogous to the return value of componentContains.  This may be necessary later.